source/Octopus.Tentacle/Kubernetes/Diagnostics/KubernetesAgentMetrics.cs

```diff
@@ -1,80 +1,80 @@
 using System;
 using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
 using Newtonsoft.Json;
 using Octopus.Diagnostics;
 using Octopus.Tentacle.Util;
 
 namespace Octopus.Tentacle.Kubernetes.Diagnostics
 {
+    using MetricsDictionary = Dictionary<string, Dictionary<string, List<DateTimeOffset>>>;
     public class KubernetesAgentMetrics
     {
-        readonly string lastEventTimestampKey = "latestTimestamp";
+        const string LastEventTimestampKey = "latestTimestamp";
 
-        public delegate KubernetesAgentMetrics Factory(IPersistenceProvider persistenceProvider);
+        public delegate KubernetesAgentMetrics Factory(IPersistenceProvider persistenceProvider, string configMapKey);
 
+        readonly SemaphoreSlim semaphore = new(1);
         readonly IPersistenceProvider persistenceProvider;
+        readonly string configMapKey;
         readonly ISystemLog log;
 
-        public KubernetesAgentMetrics(IPersistenceProvider persistenceProvider, ISystemLog log)
+        public KubernetesAgentMetrics(IPersistenceProvider persistenceProvider, string configMapKey, ISystemLog log)
         {
             this.persistenceProvider = persistenceProvider;
+            this.configMapKey = configMapKey;
             this.log = log;
         }
 
-        public void TrackEvent(string reason, string source, DateTimeOffset occurrence)
+        public async Task TrackEvent(string reason, string source, DateTimeOffset occurrence, CancellationToken cancellationToken)
         {
             try
             {
-                lock (persistenceProvider)
+                using var _ = await semaphore.LockAsync(cancellationToken);
+                await PerformUpdateOnMetricsForReason(reason, reasonDictionary =>
                 {
-                    var sourceEventsForReason = LoadFromPersistence(reason) ?? new Dictionary<string, List<DateTimeOffset>>();
-
-                    if (!sourceEventsForReason.TryGetValue(source, out var occurenceTimestamps))
+                    if (!reasonDictionary.TryGetValue(source, out var occurenceTimestamps))
                     {
-                        occurenceTimestamps = new List<DateTimeOffset>();
-                        sourceEventsForReason[source] = occurenceTimestamps;
+                        reasonDictionary[source] = occurenceTimestamps = new List<DateTimeOffset>();
```

Comment:
I cannot quite reason about this line - why are there 2 assignments?
Assume it makes all 3 the same thing i.e. a new list?

Reply:
Accepted in outside discussions